File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.js
function solution(maps) {
    var n = maps.length; //행
    var m = maps[0].length; //열

    var dir = [[0,1], [1,0], [0,-1], [-1,0]]; //오,아,왼,위
    
    var bfs = () => {
        //탐색 시작지점
        var queue = [[0,0,1]];
        maps[0][0] = 0;
        
        while(queue.length > 0) {
            var [row, cal, distance] = queue.shift();
            
            //도착 여부 판단
            if (row === n-1 && cal === m-1)
                return distance;
            
            //방향에 맞추어 방문하지 않았고, 길이라면 큐에 삽입
            for (var [r, c] of dir){
                var newr = row + r;
                var newc = cal + c;
                
                if (newr >= 0 && newr < n && newc >= 0 && newc < m && (maps[newr][newc] === 1)){
                    queue.push([newr, newc, distance + 1]);
                    maps[newr][newc] = 0;
                }
            }
        }
        return -1;
    }
    
    return bfs();
}

File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.py
from collections import deque

def solution(maps):
    n = len(maps)
    m = len(maps[0])
    
    queue = deque()
    queue.append((0,0))
    
    move = [[0,1], [1,0], [0, -1], [-1, 0]]
    
    while queue:
        c_x, c_y = queue.popleft()
        
        for d_x, d_y in move:
            n_x = c_x + d_x
            n_y = c_y + d_y
            
            if 0 <= n_x < n and 0 <= n_y < m:
                if maps[n_x][n_y] == 1:
                    maps[n_x][n_y] = maps[c_x][c_y] + 1
                    queue.append((n_x, n_y))

    if maps[n-1][m-1] > 1:
        return maps[n-1][m-1]
    else:
        return -1